src/web_socket.toit

```diff
@@ -0,0 +1,405 @@
+    return result
+
+  /**
+  Sends a ByteArray or string as a framed WebSockets message.
+  Strings are sent as text, whereas byte arrays are sent as binary.
+  */
+  send data -> none:
+    writer := start_sending --size=data.size
+    writer.write data
+    writer.close
+
+  /**
+  Starts sending a message to the peer.
+  If the size is given, the message is sent as a single fragment so that
+    the receiver knows the size.  In theory, proxies could split the fragment
+    into multiple fragments, which would ruin this.  In practice this seems
+    rare, and with TLS it is almost impossible.
+  The fragment size is at the WebSockets level and has no connection with
+    the size of IP packets or TLS buffers.
+  The message is sent as a text message if the first data written to the
+    writer is a string, otherwise as a binary message.
+  Returns a writer, which must be completed (all data sent, and closed) before
+    this method can be called again.
+  */
+  start_sending --size/int?=null -> WebSocketWriter:
+    if current_writer_: throw "PREVIOUS_WRITER_NOT_CLOSED"
+    current_writer_ = WebSocketWriter.private_ this size
+    return current_writer_
+
+  writer_close_ writer/WebSocketWriter -> none:
+    if current_writer_ != writer: throw "WRONG_WRITER_CALLED_CLOSE"
+    current_writer_ = null
+
+  write_ data from=0 to=data.size -> none:
+    written := 0
+    while from < to:
+      from += socket_.write data from to
+
+  reader_close_ -> none:
+    current_reader_ = null
+
+  /**
+  Abruptly closes the WebSocket.
+  Does not send a close packet first.
+  */
+  close:
+    socket_.close
+    current_writer_ = null
+    current_reader_ = null
+
+  static add_client_upgrade_headers_ headers/Headers -> string:
+    nonce := base64.encode (ByteArray 16: random 0x100)
+    headers.add "Connection" "upgrade"
+    headers.add "Upgrade" "websocket"
+    headers.add "Sec-WebSocket-Key" nonce
+    headers.add "Sec-WebSocket-Version" "13"
+    return nonce
+
+  static check_client_upgrade_response_ response/Response nonce/string [on_error] -> none:
+    if response.status_code != STATUS_SWITCHING_PROTOCOLS:
+      on_error.call response
+      throw response.stringify
+    upgrade_header := response.headers.single "Upgrade"
+    connection_header := response.headers.single "Connection"
+    if not upgrade_header
+        or not connection_header
+        or (Headers.ascii_normalize_ upgrade_header) != "Websocket"
+        or (Headers.ascii_normalize_ connection_header) != "Upgrade"
+        or (response.headers.single "Sec-WebSocket-Accept") != (WebSocket.response_ nonce):
+      throw "MISSING_HEADER_IN_RESPONSE"
+    if response.headers.single "Sec-WebSocket-Extensions"
+        or response.headers.single "Sec-WebSocket-Protocol":
+      throw "UNKNOWN_HEADER_IN_RESPONSE"
+
+  /**
+  Checks whether the request is a WebSocket upgrade request.
+  If it is a valid upgrade request, adds the required headers to the response_writer
```

Comment:
Document what the response is if it is an upgrade request: the nonce that was given by the client.

Reply:
There's a bunch of headers that need to be right - I think the RFC documents this just fine.